src/controller/CHIPDeviceController.cpp

```diff
@@ -294,46 +296,97 @@ CHIP_ERROR ChipDeviceController::EstablishSecureSession(NodeId peer)
 
     mSessionManager->SetDelegate(this);
 
+    err = mSessionManager->NewPairing(
+        Optional<Transport::PeerAddress>::Value(Transport::PeerAddress::UDP(mDeviceAddr, mDevicePort, mInterface)),
+        mSecurePairingSession);
+    SuccessOrExit(err);
+
     mConState = kConnectionState_SecureConnected;
 
-    if (mStorageDelegate != nullptr)
+    SendCachedPackets();
+
+exit:
+
+    if (err != CHIP_NO_ERROR)
     {
-        const char * credentials;
-        const char * address;
+        if (mSessionManager != nullptr)
+        {
+            chip::Platform::Delete(mSessionManager);
+            mSessionManager = nullptr;
+        }
+        mConState = kConnectionState_NotConnected;
+    }
+
+    return err;
+}
 
-        PERSISTENT_KEY_OP(peer, kDeviceCredentialsKeyPrefix, key, credentials = mStorageDelegate->GetKeyValue(key));
-        PERSISTENT_KEY_OP(peer, kDeviceAddressKeyPrefix, key, address = mStorageDelegate->GetKeyValue(key));
+void ChipDeviceController::OnValue(const char * key, const char * value)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    NodeId peer    = mRemoteDeviceId.Value();
 
-        SecurePairingSessionSerialized serialized;
+    VerifyOrExit(key != nullptr, err = CHIP_ERROR_INVALID_ARGUMENT);
 
-        VerifyOrExit(credentials != nullptr, err = CHIP_ERROR_KEY_NOT_FOUND_FROM_PEER);
-        VerifyOrExit(strlen(credentials) <= sizeof(serialized.inner), err = CHIP_ERROR_INVALID_STRING_LENGTH);
-        strncpy(Uint8::to_char(serialized.inner), credentials, sizeof(serialized.inner));
+    PERSISTENT_KEY_OP(
+        peer, kDeviceCredentialsKeyPrefix, expectedKey, if (strcmp(key, expectedKey) == 0) {
+            SecurePairingSessionSerialized serialized;
 
-        err = pairing.Deserialize(serialized);
-        SuccessOrExit(err);
+            VerifyOrExit(value != nullptr, err = CHIP_ERROR_INVALID_ARGUMENT);
+            VerifyOrExit(strlen(value) <= sizeof(serialized.inner), err = CHIP_ERROR_INVALID_ARGUMENT);
+            strncpy(Uint8::to_char(serialized.inner), value, sizeof(serialized.inner));
```

Comment:
do we really want strncpy() here?  does memmove() get the job done without a cast?